### DISCUSSION TITLE
how define a dynamic type with optional argument ?

### DISCUSSION THREAD

**Discussion by baixiaoyu2997** [type-challenges/type-challenges#17219] [Q&A] [answered]
我想要这种效果可以实现吗？如果useTest传入了参数text，那么getXXX可以不传入参数，但是如果useTest没有传入参数text，那么getXXX必须传入参数，也就是说getXXX的参数能否为动态可选的？

```ts
const useTest=(useTestOptions?:{text?:string})=>{
     const getXXX=(options?:{text?:string})=>{
         return options.text || text
     }
     return {
         getXXX
     }
}

```

**Comment by baboon-king** [comment]
Can you trying
https://www.typescriptlang.org/play?ts=4.8.4#code/MYewdgzgLgBArhApgFUdGBeGAeZn5gAmiAZgJZiKEB8AFAIYBOA5gIwD8AXDMgJSbUYAbwCwAKBiSYoSLGaIoADWX4GLAExcYYOAFsARokb8Mg0RKmWA9FZiEQMCCF2IYUABYVmEcZYC+vlKMCnCMYDDySir0EDwwiAAeUIhEsXBEpBRUMOwwaszq3DoGRiaC9GAAnjDctGUwFZXiAWLi4jLokcqK+AgoaFC0rLxtYh1yCt3qvUio0HVAA

```ts
const useTest = <T = undefined>(arg1?: T) => {
    const getXXX = (arg2?: number) => {
        // do some things
    }
    return getXXX as T extends undefined ? (arg2: number) => any : () => any
}


const getXXX = useTest(1)

const getXXX2 = useTest()
```
![image](https://user-images.githubusercontent.com/63645337/194740266-01801cf8-789d-49de-acd5-eb14b6c92b4c.png)
![image](https://user-images.githubusercontent.com/63645337/194740279-630b6785-56a2-4713-bb10-8157f2e430ac.png)

**Reply by baixiaoyu2997** [reply]
thanks, 这确实给我提供了另外一种思路。我想参数作为常量时，这是没有问题的，但是代入到我的对象参数中，会比较复杂，等我晚些有时间了再继续研究下，谢了哥们🤗

**Reply by baboon-king** [reply]
是的，对于简单类型参数，他可以工作。
对于对象参数，可以多余的函数签名来实现 [overloaded-functions](https://www.typescriptlang.org/docs/handbook/declaration-files/by-example.html#overloaded-functions)


https://www.typescriptlang.org/play?ts=4.8.4#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgVQGdgAVYQmAeTBgQiULgG8BYAKDk7iUwFtgAXHApRkAc3Zc4qEDCEjx7AL7t26AK5JstenHXEyFABQBKIUYg06DIUVLkqV+oRNwAvAD5hMUUglsNLR0kPQMHAB4SOFBUJAATRkpeBBhwu0NHYMIAGjgAchkYPI8PCycbOBIzODKs2zCKaizXT29ff0Dta1D7Ckjo2WB4xnSHJutCUss6yurTdy8FP1UAzS7dfV6YWomAfnqt8edXVg4ubGd4MWAYAA179xrpiYAmfYIGzImWr1OpKSgN3UUBCACIAIyQ0GSLgqNgwziAmDAkLXO73ZQrC4MK43e63cGPTYZUxYy5wNH4l5Ez5GJjcPiCOCggBCLNBcCUJjJOIpePuAGYaVs6Qz+EJQQBBSWg3KFCVLMQcrkrIA
```ts
export interface UseTestOptions {
    name: string
    text: string
}

function useTest(): (options: UseTestOptions) => string
function useTest<T extends Omit<UseTestOptions, 'text'>>(options: T): (options: UseTestOptions) => string
function useTest<T extends UseTestOptions>(options: T): () => string


function useTest(options?: UseTestOptions) {
    const getXXX = (options2?: UseTestOptions) => {
        return "111"
    }

    return getXXX
}


const getXXX1 = useTest()

const getXXX2 = useTest({ name: "BB" })

const getXXX3 = useTest({ name: "AA", text: "string" })


```

**Comment by JiatLn** [comment, marked as answer]
hope it could work~

```ts
function useTest<T extends { text: string }>(options: { text: string }): { getXXX: () => string }
function useTest<T extends Omit<Record<string, any>, 'text'>>(options: any): { getXXX: (options: { text: string }) => string }
function useTest<T extends Omit<Record<string, any>, 'text'>>(): { getXXX: (options: { text: string }) => string }


function useTest(useTestOptions: { text?: string } = {}) {

  const getXXX = (options?: { text: string }) => {
    return options.text || useTestOptions.text
  }

  return {
    getXXX
  }
}

const { getXXX: f1 } = useTest({ text: '123' })
const { getXXX: f2 } = useTest()
const { getXXX: f3 } = useTest({})

f1()
f2({ 'text': '123' })
// @ts-expect-error
f2()
// @ts-expect-error
f3()
```

**Reply by baboon-king** [reply]
cool! best idea

**Reply by baixiaoyu2997** [reply]
cool cool cool ， 重载里的泛型是不是没用到
